### PR TITLE
Fix audio player volume control on Safari

### PR DESCRIPTION
**Pull request by minacle**
_\<Audio>_ is playing audio using _AudioContext_, not using _HTMLAudioElement_ directly.
Therefore, changing _volume_ and _muted_ of _HTMLAudioElement_ are not working at all.

This PR makes _\<Audio>_ to handle volume change correctly by using _GainNode_ for it.

### CODE REVIEW

**Comment by ClearlyClaire** [comment]
Hi! Quickly looking over the change, it seems fine, but I am unable to replicate the issue it aims to fix. Which browser did you encounter the issue with?

**Comment by minacle** [comment]
@ClearlyClaire 
> Which browser did you encounter the issue with?

I am using Safari 16.2 on macOS 13.1 (22C65). I think the issue will be replicable on iOS or iPadOS, but I did not tested on these OSes.

**Review by Gargron — APPROVED** [review]

**Comment by minacle** [comment]
I've verified that existing code works fine in other major browsers.
And so, I have changed the code to make the experience consistent across all browsers.

Now fixes an issue where the effect of Gain may be applied twice (because of volume of _HTMLAudioElement_)

File: app/javascript/mastodon/features/audio/index.js [hunk 1]
```diff
@@ -336,11 +336,16 @@ class Audio extends React.PureComponent {
     const AudioContext = window.AudioContext || window.webkitAudioContext;
     const context      = new AudioContext();
     const source       = context.createMediaElementSource(this.audio);
+    const gainNode     = context.createGain();
+
+    gainNode.gain.value = this.audio.volume;
```
Comment (ClearlyClaire) [inline app/javascript/mastodon/features/audio/index.js:341]: This doesn't sound right, if `gainNode.gain.value` is supposed to replace `this.audio.volume` and `this.audio.volume` can be applied on top of `gainNode.gain.value`.

There are also still references to `this.audio.muted` on lines 84, 118 and 327, as well as references to `this.audio.volume` on lines 83, 118, and 323.

Comment (minacle) [reply app/javascript/mastodon/features/audio/index.js]: Thank you so much, @ClearlyClaire. You perfectly pointed out what I was missing.

File: app/javascript/mastodon/features/audio/index.js [hunk 2]
```diff
@@ -313,19 +309,14 @@ class Audio extends React.PureComponent {
...
 
-    if (volume !== undefined) {
-      this.audio.volume = volume;
-    }
-
-    if (muted !== undefined) {
-      this.audio.muted = muted;
-    }
+    this.audio.volume = 1;
+    this.audio.muted = false;
```
Comment (minacle) [inline app/javascript/mastodon/features/audio/index.js:319]: I'm not sure if I should keep this part or if this is the right place.

Comment (ClearlyClaire) [reply app/javascript/mastodon/features/audio/index.js:319]: That's a good question. I think it should rather be in `setAudioRef`.

File: app/javascript/mastodon/features/audio/index.js [hunk 3]
```diff
@@ -336,11 +329,16 @@ class Audio extends React.PureComponent {
     const AudioContext = window.AudioContext || window.webkitAudioContext;
     const context      = new AudioContext();
     const source       = context.createMediaElementSource(this.audio);
+    const gainNode     = context.createGain();
+
+    gainNode.gain.value = this.state.volume;
```
Comment (ClearlyClaire) [inline app/javascript/mastodon/features/audio/index.js:334]: Shouldn't this also depend on `this.state.muted`?

```suggestion
    gainNode.gain.value = this.state.muted ? 0 : this.state.volume;
```

File: app/javascript/mastodon/features/audio/index.js [hunk 4]
```diff
@@ -280,7 +281,7 @@ class Audio extends React.PureComponent {
 
     if(!isNaN(x)) {
       this.setState({ volume: x }, () => {
-        this.audio.volume = x;
+        this.gainNode.gain.value = this.state.muted ? 0 : x;
```
Comment (ClearlyClaire) [inline app/javascript/mastodon/features/audio/index.js:284]: Sorry I missed that, but it needs to check that `this.gainNode` is defined, which might not be the case if the user attempts to change the volume before starting playback.

```suggestion
        if (this.gainNode) {
          this.gainNode.gain.value = this.state.muted ? 0 : x;
        }
```

File: app/javascript/mastodon/features/audio/index.js [hunk 5]
```diff
@@ -202,7 +203,7 @@ class Audio extends React.PureComponent {
     const muted = !this.state.muted;
 
     this.setState({ muted }, () => {
-      this.audio.muted = muted;
+      this.gainNode.gain.value = muted ? 0 : this.state.volume;
```
Comment (ClearlyClaire) [inline app/javascript/mastodon/features/audio/index.js:206]: Sorry I missed that, but it needs to check that this.gainNode is defined, which might not be the case if the user attempts to change the volume before starting playback.

```suggestion
      if (this.gainNode) {
        this.gainNode.gain.value = muted ? 0 : this.state.volume;
      }
```

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]

**Review by ClearlyClaire — APPROVED** [review]
Looks good, thank you for your contribution!

**Pull request merged**